### PR TITLE
Added ROI extraction and Analysis

### PR DESCRIPTION
**Pull request by JuliGH6**
Multiple Changes were made:

- added Section to ReadMe to explain how to run the ROI extraction and analysis script
- Scripts added: run_roi (runs simple ROI extraction and analysis, custom_evaluation (writes excel files for different parameters)
- Util added: preProcessing.ipynb and extractImageInformation.ipynb are files I used to visualize single steps of the image processing. file_conversion.py includes functions which convert the file from n5 to tiff and apply masking and downsampling. excel_output.py are functions that help me create excel files from the results. extract_roi.py defines all the functions needed to get the ROIs.
- In main_qa.py I added a run_roi function which run the roi extraction and analysis
- in small metrics I changed NCC metrics to be positive, mutual information and normalized mutual information are corrected

### CODE REVIEW

File: scripts/custom_evaluation.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    # default_config["image_2"] = '/scratch/cortical_s0_cropped_downsampled.tif'
+
+
+
+
+    # default_config["transform_matrix"] = [
+    #     [1, 0, 0, 0],  # Z
+    #     [0, 1, 0, 0],  # Y
+    #     [0, 0, 1, 0],  # X
+    #     [0, 0, 0, 1],
```
Comment (camilolaiton) [inline scripts/custom_evaluation.py:57]: remove the comments if not necessary

File: scripts/custom_evaluation.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+    window_sizes = [30]
+    num_of_points = [60]
+
+    # corticals = ['/scratch/cortical_s0_cropped_downsampled_f1.tif', '/scratch/cortical_s0_cropped_downsampled_f2.tif']
+    # confocals = ['/scratch/confocal_s0_cropped_downsampled_f1.tif', '/scratch/confocal_s0_cropped_downsampled_f2.tif']
+    corticals = ['/scratch/img1.tif']
+    confocals = ['/scratch/img2.tif']
+
+    matrizes = [
```
Comment (camilolaiton) [inline scripts/custom_evaluation.py:75]: are these transforms the ones used for evaluation? If so, I would recommend adding a parameter that will receive the transforms so they are not hardcoded, please do this as well with the images.

File: src/aind_registration_evaluation/metric/small_scale.py [hunk 3]
```diff
@@ -434,28 +421,14 @@ def normalized_mutual_information(
...
-        mutual_information = numerator / denominator
+        # Check performance of float64 in large-scale datasets
+        # since converting to float64 required way more memory.
+        patch_1 = patch_1.flatten()  # .astype(np.float64)
+        patch_2 = patch_2.flatten()  # .astype(np.float64)
 
-        return mutual_information
+        # Compute the Mutual Information between the two image pixel distributions
+        # using skimage
+        return normalized_mutual_info_score(patch_1, patch_2, average_method='geometric')
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/metric/small_scale.py:431]: Please, add this in the LargeScale class.

File: src/aind_registration_evaluation/metric/small_scale.py [hunk 4]
```diff
@@ -388,24 +390,9 @@ def mutual_information(
...
-                pxy[non_zero_pxy_pos] / px_py[non_zero_pxy_pos],
-                dtype=self.dtype,
-            ),
-            dtype=self.dtype,
-        )
-
-        return mutual_information
+        patch_1 = patch_1.flatten() 
+        patch_2 = patch_2.flatten() 
+        return mutual_info_score(patch_1, patch_2)
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/metric/small_scale.py:395]: Please, add this to the LargeScale class.

File: src/aind_registration_evaluation/util/cellpose_util.ipynb [hunk 5]
```diff
@@ -0,0 +1,617 @@
+{
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/cellpose_util.ipynb:1]: This notebook must not be please here, it is necessary or were they just experiments?

File: src/aind_registration_evaluation/util/excelOutput.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+import os
+import numpy as np
+import pandas as pd
+
+def create_excel_file(file_path):
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/excelOutput.py:5]: There utilities could be added to the utils package. Please make sure for new files you follow the **file_name.py** convention.

File: src/aind_registration_evaluation/util/extractImageInformation.ipynb [hunk 7]
```diff
@@ -0,0 +1,233 @@
+{
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/extractImageInformation.ipynb:1]: Same as above with the notebook comment.

File: src/aind_registration_evaluation/util/extract_roi.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
+import numpy as np
+from skimage import io, filters, morphology, exposure, measure
+import matplotlib.pyplot as plt
+import tifffile as tiff
+from sklearn.metrics import mutual_info_score, normalized_mutual_info_score
+from scipy.stats import entropy
+
+def get_props(image, binaryThreshold):
+    '''
+    	- Applies Gaussian blur to smooth the image
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/extract_roi.py:10]: Please, make sure all the documentation follows the numpy style docstrings structure.

File: src/aind_registration_evaluation/util/extract_roi.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+    blurred_image = filters.gaussian(image, sigma=1)
+    binary_image = blurred_image > binaryThreshold
+    opened_image = morphology.opening(binary_image, morphology.ball(1))
+    largeObjects = morphology.remove_small_objects(opened_image, 500)
+    labels = morphology.label(largeObjects)
+    regions = measure.regionprops(labels)
+    return regions
+
+def get_centroid_distances(props1, props2, distanceThreshold):
+    '''
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/extract_roi.py:35]: Same here, please change to numpy style docstring

File: src/aind_registration_evaluation/util/extract_roi.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+        y_min = 0 if min(bbox1[1],bbox2[1])<4 else (min(bbox1[1],bbox2[1])-3)
+        x_min = 0 if min(bbox1[2],bbox2[2])<4 else (min(bbox1[2],bbox2[2])-3)
+        z_max = image_shape[0] if max(bbox1[3],bbox2[3])>(image_shape[0]-4) else (max(bbox1[3],bbox2[3])+3)
+        y_max = image_shape[1] if max(bbox1[4],bbox2[4])>(image_shape[1]-4) else (max(bbox1[4],bbox2[4])+3)
+        x_max = image_shape[2] if max(bbox1[5],bbox2[5])>(image_shape[2]-4) else (max(bbox1[5],bbox2[5])+3)
+        patches.append((z_min, y_min, x_min, z_max, y_max, x_max))
+    
+    merged_patches = []
+    while patches:
+        z_min, y_min, x_min, z_max, y_max, x_max = patches.pop(0)
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/extract_roi.py:107]: This block of code is super dense, please break. Also, the while must be used only when you don't know how many times you will execute the loop. In this case, I see patches is a list which has a finite number of elements that you're adding.

Comment (camilolaiton) [reply src/aind_registration_evaluation/util/extract_roi.py:107]: It looks it is some optimization you're doing? If so, please ignore my comment. But break the code, or explain it a little better :)

File: src/aind_registration_evaluation/util/extract_roi.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+        
+        overlapThreshold: volume overlap theshold above which two patches are merged to their max coordinates (float)
+
+        Returns
+        ------------------------
+        dictionary of {patch_coordinates: patch_volume}
+    '''
+    img1_props = get_props(img1, img1_binaryThreshold)
+    img2_props = get_props(img2, img2_binaryThreshold)
+    print("ConfocalProps:", len(img1_props), "CorticalProps:", len(img2_props))
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/extract_roi.py:146]: remove tailed prints

File: src/aind_registration_evaluation/util/extract_roi.py [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+    distances = get_centroid_distances(img1_props, img2_props, maxCentroidDistance)
+    print("Cell Matches:", len(distances))
+    patches = create_patches(distances, img1.shape, overlapThreshold)
+    patch_dict = {r: ((r[3]-r[0]) * (r[4]-r[1]) * (r[5]-r[2])) for r in patches}
+    return patch_dict
+
+def get_ROIs_cellpose(img1, img2, maxCentroidDistance, overlapThreshold):
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/extract_roi.py:153]: this function and the one above seem pretty similar, please, optimize if possible

Comment (camilolaiton) [reply src/aind_registration_evaluation/util/extract_roi.py:153]: A helper function could be useful here

File: src/aind_registration_evaluation/util/extract_roi.py [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+    print("ConfocalProps:", len(img1_regions), "CorticalProps:", len(img2_regions))
+    cp_centroid_dist = get_centroid_distances(img1_regions, img2_regions, maxCentroidDistance)
+    print("Cell Matches:", len(cp_centroid_dist))
+    cp_patches = create_patches(cp_centroid_dist, img1.shape, 0.3)
+    patch_dict = {r: ((r[3]-r[0]) * (r[4]-r[1]) * (r[5]-r[2])) for r in cp_patches}
+    return patch_dict
+
+
+
+def normalized_mutual_information(patch_1, patch_2
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/extract_roi.py:180]: These metrics must be included in the metrics.py file

File: src/aind_registration_evaluation/util/file_conversion.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+        Returns
+        ------------------------
+        none
+
+    '''
+    f = z5py.File(n5_filename)
+    image = np.array(f["c0"])
+    tifffile.imwrite(output_path, image)
+
+def apply_mask(inFilepath, outFilepath, threshhold, newValue):
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/file_conversion.py:118]: the file is named file_conversion.py but some processing functions are included here. Please, move the function where it's convenient.

File: src/aind_registration_evaluation/util/file_conversion.py [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+
+    '''
+    data = tifffile.imread(inFilepath)
+    mask = data <= threshhold
+    data[mask] = newValue
+    tifffile.imwrite(outFilepath, data)
+
+
+def apply_mask_and_crop(inFilepath_cortical, outFilepath_cortical, inFilepath_confocal, outFilepath_confocal, threshold):
+    '''
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/file_conversion.py:141]: Same here

File: src/aind_registration_evaluation/util/file_conversion.py [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+        outFilepath_cortical: filepath for desired filelocation of cropped confocal image
+        threshold: intensity threshold under which the pixel will interpreted as not overlapping
+
+        Returns
+        ------------------------
+        none
+
+    '''
+    image = tifffile.imread(inFilepath)
+    normalized_image = cv2.normalize(image, None, 0, 255, cv2.NORM_MINMAX)
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/file_conversion.py:231]: What if you have a uint16 image? It'll fail since you're only looking in the uint8 numerical space.

File: src/aind_registration_evaluation/util/file_conversion.py [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+    :param current_resolution: Tuple of current resolutions (z_res, y_res, x_res)
+    :param desired_resolution: Desired resolution (z_res, y_res, x_res)
+    :return: Resampled 3D numpy array
+    """
+    current_z_res, current_y_res, current_x_res = current_resolution
+    desired_z_res, desired_y_res, desired_x_res = desired_resolution
+    
+    # Calculate the zoom factors for each dimension
+    zoom_factors = (
+        current_z_res / desired_z_res,
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/file_conversion.py:249]: make sure the zoom factors are nonzero

File: src/aind_registration_evaluation/util/preProcessing.ipynb [hunk 18]
```diff
@@ -0,0 +1,334 @@
+{
```
Comment (camilolaiton) [inline src/aind_registration_evaluation/util/preProcessing.ipynb:1]: same as above in terms of notebooks

**Review by camilolaiton — CHANGES_REQUESTED** [review]
A couple things to fix, also please run linters to comply with code standards
